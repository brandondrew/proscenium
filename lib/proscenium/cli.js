import { writeAll } from 'https://deno.land/std@0.129.0/streams/conversion.ts'
import buildJS from './cli/js_builder.js'

if (import.meta.main) {
  await writeAll(Deno.stdout, await init(Deno.args))
}

export async function init(args, options = {}) {
  return await buildJS(args, options)
}
