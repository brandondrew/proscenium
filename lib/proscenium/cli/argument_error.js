export default class CliArgumentError extends Error {
  static MESSAGES = {
    cwdRequired: 'Current working directory is required as first argument.',
    entrypointRequired: 'An entry point is required as second argument.',
    builderRequired: 'The builder is required as third and final argument.',

    cwdUnknown: ({ cwd }) => `A valid working directory is required - received ${cwd}`,
    entrypointUnknown: ({ entrypoint }) =>
      `A valid entrypoint is required - received ${entrypoint}`,
    builderUnknown: ({ builder }) => `Unknown builder '${builder}'`
  }

  constructor(reason, options) {
    let message = CliArgumentError.MESSAGES[reason]
    if (typeof message === 'function') {
      message = message(options)
    }

    super(message, options)

    this.reason = reason
    this.message = message
  }
}
