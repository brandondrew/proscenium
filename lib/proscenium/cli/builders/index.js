export const builderNames = ['javascript', 'react', 'solid']
