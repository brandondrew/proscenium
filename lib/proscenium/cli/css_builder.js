import { parse } from 'https://deno.land/std/flags/mod.ts'
import init, { transform } from 'https://unpkg.com/@parcel/css-wasm'

export default async function (args) {
  const { _: entrypoints, ...flags } = parse(args, {
    default: {
      minify: false,
      'css-modules': false,
      'source-map': false,
      nesting: false
    }
  })

  const entrypoint = entrypoints[0]
  const contents = await Deno.readTextFile(entrypoint)

  await init()

  const result = transform({
    filename: entrypoint,
    code: new TextEncoder().encode(contents),
    minify: flags.minify,
    drafts: {
      nesting: flags.nesting
    },
    targets: {
      safari: (13 << 16) | (2 << 8)
    },
    cssModules: flags['css-modules'],
    sourceMap: flags['source-map']
  })

  // console.log(new TextDecoder().decode(code))

  return result.code
}
