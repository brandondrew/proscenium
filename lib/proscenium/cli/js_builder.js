import * as esbuild from 'https://deno.land/x/esbuild@v0.14.8/mod.js'
import { parse } from 'https://deno.land/std/flags/mod.ts'
import { join } from 'https://deno.land/std/path/mod.ts'

import resolvePlugin from './esbuild/resolve_plugin.js'

const isProd = Deno.env.get('RAILS_ENV') === 'production'
const template = `
<style>
:host {
  position: fixed;
  z-index: 99999;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow-y: scroll;
  margin: 0;
  background: rgba(0, 0, 0, 0.66);
  display: flex;
  align-items: center;
  --monospace: 'SFMono-Regular', Consolas,
              'Liberation Mono', Menlo, Courier, monospace;
  --red: #ff5555;
  --yellow: #e2aa53;
  --purple: #cfa4ff;
  --cyan: #2dd9da;
  --dim: #c9c9c9;
}

.window {
  font-family: var(--monospace);
  line-height: 1.5;
  width: 800px;
  color: #d8d8d8;
  margin: 30px auto;
  padding: 25px 40px;
  position: relative;
  background: #181818;
  border-radius: 6px 6px 8px 8px;
  box-shadow: 0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22);
  overflow: hidden;
  border-top: 8px solid var(--red);
}

pre {
  font-family: var(--monospace);
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 1em;
  overflow-x: scroll;
  scrollbar-width: none;
}

pre::-webkit-scrollbar {
  display: none;
}

.message {
  line-height: 1.3;
  font-weight: 600;
  white-space: pre-wrap;
}

.message-body {
  color: var(--red);
}

.file {
  color: var(--cyan);
  margin-bottom: 0;
  white-space: pre-wrap;
  word-break: break-all;
}

</style>
<div class="window">
  <pre class="message"><span class="message-body"></span></pre>
  <pre class="file"></pre>
</div>
`

export default async args => {
  const { _: entrypoints, ...flags } = parse(args)
  const [cwd, entrypoint] = validatePaths(entrypoints)

  const params = {
    entryPoints: [entrypoint],
    absWorkingDir: cwd,
    logLevel: 'silent',
    write: false,
    format: 'esm',
    bundle: true,
    plugins: [resolvePlugin()]
  }

  if (entrypoint.endsWith('.jsx')) {
    try {
      const stat = Deno.lstatSync(join(cwd, 'lib/react_shim.js'))
      if (stat.isFile) {
        params.inject = ['./lib/react_shim.js']
      }
    } catch {
      // Safe to swallow as this should only throw if file does not exist.
    }
  }

  try {
    const result = await esbuild.build(params)
    return result.outputFiles[0].contents
  } catch (e) {
    if (isProd) {
      return new TextEncoder().encode(`
        const err = ${JSON.stringify(e.errors[0])}
        const location = \`\${err.location.file}:\${err.location.line}:\${err.location.column}\`
        console.error('%s at %O', err.text, location);
      `)
    } else {
      return new TextEncoder().encode(`
        class ErrorOverlay extends HTMLElement {
          constructor(err) {
            super()

            this.root = this.attachShadow({ mode: 'open' })
            this.root.innerHTML = \`${template}\`
            this.root.querySelector('.message-body').textContent = err.text.trim()

            if (err.location) {
              const location = \`\${err.location.file}:\${err.location.line}:\${err.location.column}\`
              this.root.querySelector('.file').textContent = location
              console.error('%s at %O', err.text, location)
            } else {
              console.error(err.text)
            }
          }
        }

        customElements.define('error-overlay', ErrorOverlay)
        document.body.appendChild(new ErrorOverlay(${JSON.stringify(e.errors[0])}))
      `)
    }
  } finally {
    esbuild.stop()
  }
}

function validatePaths(paths) {
  const cwd = paths[0]
  const entrypoint = paths[1]

  if (!cwd || !entrypoint) {
    throw new TypeError(
      'Current working directory and entrypoint are required first and second arguments.'
    )
  }

  try {
    const stat = Deno.lstatSync(cwd)
    if (!stat.isDirectory) {
      throw new TypeError(
        `Current working directory is required as the first argument - received ${cwd}`
      )
    }
  } catch {
    throw new TypeError(
      `A valid working directory is required as the first argument - received ${cwd}`
    )
  }

  try {
    const stat = Deno.lstatSync(join(cwd, entrypoint))
    if (!stat.isFile) {
      throw new TypeError(`Entrypoint is required as the second argument - received ${entrypoint}`)
    }
  } catch {
    throw new TypeError(
      `A valid entrypoint is required as the second argument - received ${entrypoint}`
    )
  }

  if (/\.(js|jsx)$/.test(entrypoint) === false) {
    throw new TypeError(
      `Only a JS/JSX entrypoint is supported with this CLI - received ${entrypoint}`
    )
  }

  return [cwd, entrypoint]
}
