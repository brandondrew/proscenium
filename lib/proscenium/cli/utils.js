import { join } from 'std/path/mod.ts'

import CliArgumentError from './argument_error.js'
import { builderNames } from './builders/index.js'

export const isTest = () => Deno.env.get('ENVIRONMENT') === 'test'

export const debug = (...args) => {
  isTest() && console.log(...args)
}

export const setup = (pluginName, pluginFn) => {
  return (options = {}) => ({
    name: pluginName,
    setup(build) {
      const plugin = pluginFn(build)

      if (plugin.onResolve) {
        const { callback, ...onResolve } = plugin.onResolve

        build.onResolve(onResolve, async params => {
          if (params.pluginData?.isResolvingPath) return

          const results = await callback(params)

          options.debug && console.debug(`plugin(${pluginName}:onResolve)`, { params, results })

          return results
        })
      }

      if (plugin.onLoad) {
        const { callback, ...onLoad } = plugin.onLoad

        build.onLoad(onLoad, params => {
          const results = callback(params)

          options.debug && console.debug(`plugin(${pluginName}:onLoad)`, { params, results })

          return results
        })
      }
    }
  })
}

export const parseArgs = args => {
  let [cwd, entrypoint, builder] = args

  if (!cwd) {
    throw new CliArgumentError('cwdRequired')
  }

  if (!entrypoint) {
    throw new CliArgumentError('entrypointRequired')
  }

  if (!builder) {
    throw new CliArgumentError('builderRequired')
  }

  try {
    const stat = Deno.lstatSync(cwd)
    if (!stat.isDirectory) {
      throw new CliArgumentError(
        `Current working directory is required as the first argument - received ${cwd}`
      )
    }
  } catch {
    throw new CliArgumentError('cwdUnknown', { cwd })
  }

  if (/\.(jsx?)|(css)\.map$/.test(entrypoint)) {
    entrypoint = entrypoint.replace(/\.map$/, '')
  }

  try {
    const stat = Deno.lstatSync(join(cwd, entrypoint))
    if (!stat.isFile) {
      throw new CliArgumentError(
        `Entrypoint is required as the second argument - received ${entrypoint}`
      )
    }
  } catch {
    throw new CliArgumentError('entrypointUnknown', { entrypoint })
  }

  if (!builderNames.includes(builder)) {
    throw new CliArgumentError('builderUnknown', { builder })
  }

  return args
}
