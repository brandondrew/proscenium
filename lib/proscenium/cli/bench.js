import cli from '../cli.js'

Deno.bench('react', async () => {
  await cli(Deno.args)
})
