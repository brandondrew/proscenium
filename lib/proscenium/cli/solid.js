import { writeAll } from 'std/streams/mod.ts'
import { parseArgs } from './utils.js'

import builder from './builders/solid.js'

if (import.meta.main) {
  await writeAll(Deno.stdout, await main(Deno.args))
}

async function main(args = []) {
  const [cwd, entrypoint, _] = parseArgs(args)
  return await builder(cwd, entrypoint)
}

export default main
