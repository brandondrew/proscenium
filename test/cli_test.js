import { assertRejects, assertStringIncludes } from 'https://deno.land/std/testing/asserts.ts'
import { join } from 'https://deno.land/std/path/mod.ts'

import { init } from '../lib/proscenium/cli.js'

Deno.env.set('RAILS_ENV', 'test')
Deno.env.set('PROSCENIUM_TEST', 'test')
const cwd = join(Deno.cwd(), 'test', 'internal')

Deno.test('throws without any arguments', () => {
  assertRejects(() => init(), TypeError)
})

Deno.test('throws without `cwd` argument', () => {
  assertRejects(() => init('/foo/bar'), TypeError)
})

Deno.test('throws without `entrypoint` argument', () => {
  assertRejects(() => init(['/foo/bar', 'comp.js']), TypeError)
})

Deno.test('Throws on unknown cwd', () => {
  assertRejects(() => init(['/foo/bar', 'comp.js']), TypeError)
})

Deno.test('Throws on unknown entrypoint', () => {
  assertRejects(() => init([cwd, 'comp.js']), TypeError)
})

Deno.test('Successful JS build', async () => {
  const result = await init([cwd, 'app/views/layouts/application.js'])

  assertStringIncludes(
    new TextDecoder().decode(result),
    'console.log("app/views/layouts/application.js")'
  )
})

Deno.test('Import css from JS', async () => {
  const result = await init([cwd, 'lib/import_css.js'])

  assertStringIncludes(
    new TextDecoder().decode(result),
    'ele.setAttribute("href", "/app/views/layouts/application.css");'
  )
})

Deno.test('Import css from jsx', async () => {
  const result = await init([cwd, 'lib/import_css.jsx'], { debug: true })

  assertStringIncludes(
    new TextDecoder().decode(result),
    'ele.setAttribute("href", "/app/views/layouts/application.css");'
  )
})

Deno.test('Import css module', { only: true }, async () => {
  // const result = await init([cwd, 'lib/import_css_module.js'], { debug: true })
  const result = await init([cwd, 'lib/import_assert_css.js'], { debug: true })

  assertStringIncludes(
    new TextDecoder().decode(result),
    'ele.setAttribute("href", "/lib/css_module.css");'
  )
  assertStringIncludes(new TextDecoder().decode(result), '{ myClass: "123_myClass" }')
})
