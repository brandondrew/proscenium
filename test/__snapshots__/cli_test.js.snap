export const snapshot = {};

snapshot[`cli > import map 1`] = `
'// lib/import_map.js
import axios from "https://cdnjs.cloudflare.com/ajax/libs/axios/0.24.0/axios.min.js";
import foo from "/lib/foo.js";
axios();
foo();
import("https://esm.sh/react@18.1.0");
'
`;

snapshot[`cli > Import css module from JS 1`] = `
'// importStylesheet:/lib/styles.module.css
import { importCssModule } from "/proscenium-runtime/import_css.js";
var styles_module_default = await importCssModule("/lib/styles.module.css");

// lib/import_css_module.js
console.log(styles_module_default);
'
`;

snapshot[`cli > Import css from JS 1`] = `
'// importStylesheet:/app/views/layouts/application.css
import { appendStylesheet } from "/proscenium-runtime/import_css.js";
appendStylesheet("/app/views/layouts/application.css");
'
`;

snapshot[`cli > Import css from jsx 1`] = `
'// importStylesheet:/app/views/layouts/application.css
import { appendStylesheet } from "/proscenium-runtime/import_css.js";
appendStylesheet("/app/views/layouts/application.css");
'
`;
